src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1637,7 +1591,7 @@ internal BinderFactory GetBinderFactory(SyntaxTree syntaxTree)
             var binderFactories = _binderFactories;
             if (binderFactories == null)
             {
-                binderFactories = new WeakReference<BinderFactory>[_syntaxTrees.Length];
+                binderFactories = new WeakReference<BinderFactory>[_syntaxAndDeclarations.GetLazyState().SyntaxTrees.Length];
```

Comment:
Consider using SyntaxTrees rather than _syntaxAndDeclarations.GetLazyState().SyntaxTrees, here and elsewhere.